### PR TITLE
Make `MedianOfRunningArray` Generic

### LINKED ISSUE #4390: [FEATURE REQUEST] Make `MedianOfRunningArray` generic
### What would you like to Propose?

I find it a bit odd that the [`MedianOfRunningArray`](https://github.com/TheAlgorithms/Java/blob/fbe348b105a06258a81538023698117e2bb116cb/src/main/java/com/thealgorithms/misc/MedianOfRunningArray.java) only works with integers and returns double as the median. I suggest to make `MedianOfRunningArray` generic accepting any [`Number`](https://docs.oracle.com/javase/8/docs/api/java/lang/Number.html) class.

### Issue details

I suggest to define `MedianOfRunningArray` as

```java
public class MedianOfRunningArray<T extends Number> {
    //...
}
```

and to change the method `insert` to accept objects of `T` and `median` to return `T`.

### Additional Information

Please make sure to include the following test (_rounding down to the integer_):

```Java
var stream = new MedianOfRunningArray<Integer>();
stream.insert(1);
stream.insert(2);
assertEquals(1, stream.median());
```

### PR DESCRIPTION
**Pull request by govardhanshah456**
Closes #4390

### CODE REVIEW

**Comment by govardhanshah456** [comment]
@vil02 Hey, i did some changes according to my knowledge of generics. You can check it.

File: src/main/java/com/thealgorithms/misc/MedianOfRunningArray.java [hunk 1]
```diff
@@ -38,14 +38,40 @@ public void insert(Integer e) {
...
+            int sum = ((Short) a) + ((Short) b);
+            return (T) Short.valueOf((short) (sum / 2));
+        } else if (a instanceof Byte) {
+            int sum = ((Byte) a) + ((Byte) b);
+            return (T) Byte.valueOf((byte) (sum / 2));
+        } else {
+            throw new IllegalArgumentException("Unsupported numeric type");
         }
-        return maxHeap.size() > minHeap.size() ? maxHeap.peek() * 1.0 : minHeap.peek() * 1.0;
     }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MedianOfRunningArray.java:76]: This is ugly. I think that it is fair to say that generics in Java are annoying. What do you think about defining the `MedianOfRunningArray` as abstract class (with `computeAverage` as an abstract method) and then have classes like `MedianOfRunningArrayInteger`, `MedianOfRunningArrayDouble` extending it?

In more details (I made some slight changes, please consider them):

```java
public abstract class MedianOfRunningArray<T extends Number & Comparable<T>> {

    private PriorityQueue<T> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
    private PriorityQueue<T> minHeap = new PriorityQueue<>();

    /*
      Inserting lower half of array to max Heap
      and upper half to min heap
     */
    public void insert(final T e) {
        if (!minHeap.isEmpty() && e.compareTo(minHeap.peek()) < 0) {
            maxHeap.offer(e);
            if (maxHeap.size() > minHeap.size() + 1) {
                minHeap.offer(maxHeap.poll());
            }
        } else {
            minHeap.offer(e);
            if (minHeap.size() > maxHeap.size() + 1) {
                maxHeap.offer(minHeap.poll());
            }
        }
    }

    /*
      Returns median at any given point
     */
    public T median() {
        if (maxHeap.isEmpty() && minHeap.isEmpty()) {
            throw new IllegalArgumentException("Enter at least 1 element, Median of empty list is not defined!");
        } else if (maxHeap.size() == minHeap.size()) {
            T maxHeapTop = maxHeap.peek();
            T minHeapTop = minHeap.peek();
            return calculateAverage(maxHeapTop, minHeapTop);
        }
        return maxHeap.size() > minHeap.size() ? maxHeap.peek() : minHeap.peek();
    }

    public abstract T calculateAverage(T a, T b);
}
```

and

```java
final public class MedianOfRunningArrayDouble extends MedianOfRunningArray<Double> {
    @Override
    public Double calculateAverage(final Double a, final Double b) {
        return (a + b) / 2.0d;
    }
}
```

```java
final public class MedianOfRunningArrayInteger extends MedianOfRunningArray<Integer> {
    @Override
    public Integer calculateAverage(final Integer a, final Integer b) {
        return (a + b) / 2;
    }
}
```

Having `calculateAverage` as public will help with testing.

The advantage of this approach is that when someone wants to have a different rounding for integers it is enough to have `MedianOfRunningArrayIntegerCrazyRounding`.

File: src/test/java/com/thealgorithms/misc/MedianOfRunningArrayTest.java [hunk 2]
```diff
@@ -17,71 +17,71 @@ public class MedianOfRunningArrayTest {
```
Comment (vil02) [inline src/test/java/com/thealgorithms/misc/MedianOfRunningArrayTest.java:17]: Please clean-up the imports and the comment/documentation. I forgot to write that in the previous PR. Sorry.

**Review by vil02 — CHANGES_REQUESTED** [review]

**Comment by govardhanshah456** [comment]
@vil02 Hey ,as suggested I have made class abstract and did the required changes, have a look.

File: src/main/java/com/thealgorithms/misc/MedianOfRunningArrayInteger.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.thealgorithms.misc;
+
+final public class MedianOfRunningArrayInteger extends MedianOfRunningArray<Integer> {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MedianOfRunningArrayInteger.java:3]:
```suggestion
public final class MedianOfRunningArrayInteger extends MedianOfRunningArray<Integer> {
```

File: src/main/java/com/thealgorithms/misc/MedianOfRunningArrayLong.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.thealgorithms.misc;
+
+final public class MedianOfRunningArrayLong extends MedianOfRunningArray<Long> {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MedianOfRunningArrayLong.java:3]:
```suggestion
public final class MedianOfRunningArrayLong extends MedianOfRunningArray<Long> {
```

File: src/test/java/com/thealgorithms/misc/MedianOfRunningArrayTest.java (hunk 2 above)
Comment (vil02) [reply src/test/java/com/thealgorithms/misc/MedianOfRunningArrayTest.java:17]: `import static org.junit.jupiter.api.Assertions.assertFalse;` and `import org.apache.commons.lang3.tuple.Pair;` are not needed.

The comments
```
/**
 * Test case for Two sum Problem.
 * @author Bama Charan Chhandogi (https://github.com/BamaCharanChhandogi)
 */
```
seems to be wrong.

**Review by vil02 — CHANGES_REQUESTED** [review]
It goes in a very good direction!

**Comment by govardhanshah456** [comment]
@vil02 Hey, modified that comments and removed that impot and committed suggested changes, regarding double  and float i did not know that there exist some way to check precision using assertEqual itself, so was formatting at output side.

**Review by vil02 — DISMISSED** [review]
Looks good. Thanks!

@siriak could you please have a look? There were _lots_ of new files introduced and I would like to double check if the current structure is correct.

File: src/main/java/com/thealgorithms/misc/MedianOfRunningArrayFloat.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.thealgorithms.misc;
+public final class MedianOfRunningArrayFloat extends MedianOfRunningArray<Float> {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MedianOfRunningArrayFloat.java:2]:
```suggestion

public final class MedianOfRunningArrayFloat extends MedianOfRunningArray<Float> {
```

**Review by vil02 — APPROVED** [review]

**Comment by govardhanshah456** [comment]
Yeah, thanks.

**Review by siriak — APPROVED** [review]
Looks good, thanks!

**Comment by vil02** [comment]
@govardhanshah456 this was a good piece of Java programing. Thanks!

**Pull request merged**